cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -0,0 +1,295 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cmath>
+#include <cassert>
+#include <limits>
+#include <functional>
+#include <boost/serialization/strong_typedef.hpp>
+
+namespace cudf {
+namespace fp {
+
+BOOST_STRONG_TYPEDEF(int32_t, scale_type)
+
+enum Radix : int32_t {
+    BASE_2  = 2,
+    BASE_10 = 10
+};
+
+namespace detail {
+    // helper function to negate strongly typed scale_type
+    auto negate(scale_type const& scale) -> scale_type{
+        return scale_type{-1 * scale};
+    }
+
+    // perform this operation when constructing with - scale (after negating scale)
+    template <Radix Rad, typename T>
+    constexpr auto right_shift(T const& val, scale_type const& scale) {
+        return val / std::pow(static_cast<int32_t>(Rad), static_cast<int32_t>(scale));
+    }
+
+    // perform this operation when constructing with + scale
+    template <Radix Rad, typename T>
+    constexpr auto left_shift(T const& val, scale_type const& scale) {
+        return val * std::pow(static_cast<int32_t>(Rad), static_cast<int32_t>(scale));
+    }
+
+    // convenience generic shift function
+    template <Radix Rad, typename T>
+    constexpr auto shift(T const& val, scale_type const& scale) {
+        if      (scale == 0) return static_cast<double>(val);
+        else if (scale >= 0) return right_shift<Rad>(val, scale);
+        else                 return left_shift <Rad>(val, negate(scale));
+    }
+}
+
+// helper struct for constructing fixed_point when value is already shifted
+template <typename Rep>
+struct scaled_integer{
+    Rep value;
+    scale_type scale;
+    explicit scaled_integer(Rep v, scale_type s) : value(v), scale(s) {}
+};
+
+
+// Rep = representative type
+template <typename Rep, Radix Rad>
+class fixed_point {
+
+    scale_type _scale;
+    Rep        _value;
+
+public:
+
+    // CONSTRUCTORS
+    template <typename T = Rep,
+              typename std::enable_if_t<(std::numeric_limits<T>::is_integer
+                                      || std::is_floating_point<T>::value)>* = nullptr>
+    explicit fixed_point(T const& value, scale_type const& scale) :
+        _value(detail::shift<Rad>(value, scale)),
+        _scale(scale)
+    {
+    }
+
+    explicit fixed_point(scaled_integer<Rep> s) :
+        _value(s.value),
+        _scale(s.scale)
+    {
+    }
+
+    fixed_point() = default; // _value & _scale = 0
+
+    // EXPLICIT CONVERSION OPERATOR
+    template <typename U,
+              typename std::enable_if_t<(std::numeric_limits<U>::is_integer
+                                      || std::is_floating_point<U>::value)>* = nullptr>
+    explicit constexpr operator U() const {
+        return detail::shift<Rad>(static_cast<U>(_value), detail::negate(_scale));
+    }
+
+    auto get() const noexcept {
+        return static_cast<double>(*this);
```

Comment:
I believe so, but currently looking into SFINAE `fixed_point` to the integer types (`int32_t`, `int64_t`, etc) that we intend to be used for `Rep`. Will update once done.

Reply:
We know have:
```cpp
template <typename T>
constexpr inline auto is_supported_representation_type() -> bool {
  return std::is_same<T, int32_t>::value;
}
```
Which the constructors used to SFINAE :+1:

It can very easily be extended in the future to support other types.